ceviche/fdfd.py

```diff
@@ -191,6 +191,123 @@ def _solve_fn(self, eps_vec, entries_a, indices_a, Jz_vec):
         Hx_vec, Hy_vec = self._Ez_to_Hx_Hy(Ez_vec)
         return Hx_vec, Hy_vec, Ez_vec
 
+
+class fdfd_mf_ez(fdfd):
+    """ FDFD class for multifrequency linear Ez polarization """
+
+    def __init__(self, omega, dL, eps_r, Omega, delta, phi, Nsb, npml, bloch_phases=None):
+        super().__init__(omega, dL, eps_r, npml, bloch_phases=bloch_phases)
+        self.Omega = Omega
+        self.delta = delta
+        self.phi = phi
+        self.Nsb = Nsb
+
+    @property
+    def delta(self):
+        """ Returns the relative permittivity grid """
+        return self._delta
+
+    @property
+    def phi(self):
+        """ Returns the relative permittivity grid """
+        return self._phi
+
+    @delta.setter
+    def delta(self, new_delta):
+        """ Defines some attributes when delta is set. """
+        self._delta = new_delta
+
+    @phi.setter
+    def phi(self, new_phi):
+        """ Defines some attributes when [hi] is set. """
+        self._phi = new_phi
+
+    def solve(self, source_z):
+        """ Outward facing function (what gets called by user) that takes a source grid and returns the field components """
+        # flatten the permittivity and source grid
+        source_vec = self._grid_to_vec(source_z)
+        eps_vec = self._grid_to_vec(self.eps_r)
+        Nfreq = npa.shape(self.delta)[0]
+        delta_matrix = self.delta.reshape([Nfreq, npa.prod(self.shape)])
+        phi_matrix = self.phi.reshape([Nfreq, npa.prod(self.shape)])
+        # create the A matrix for this polarization
+        entries_a, indices_a = self._make_A(eps_vec, delta_matrix, phi_matrix)
+
+        # solve field componets usng A and the source
+        Fx_vec, Fy_vec, Fz_vec = self._solve_fn(eps_vec, entries_a, indices_a, source_vec)
+
+        # put all field components into a tuple, convert to grid shape and return them all
+        Fx = self._vec_to_grid(Fx_vec)
+        Fy = self._vec_to_grid(Fy_vec)
+        Fz = self._vec_to_grid(Fz_vec)
+
+        return Fx, Fy, Fz
+
+    def _make_A(self, eps_vec, delta_matrix, phi_matrix):
+
+        M = 2*self.Nsb + 1
+        N = self.Nx * self.Ny
+        W = self.omega + npa.arange(-self.Nsb,self.Nsb+1)*self.Omega
+
+        C = sp.kron(sp.eye(M), - 1 / MU_0 * self.Dxf.dot(self.Dxb) - 1 / MU_0 * self.Dyf.dot(self.Dyb))
+        entries_c, indices_c = get_entries_indices(C)
+
+        # indices into the diagonal of a sparse matrix 
+        D = - EPSILON_0 * sp.kron(sp.spdiags(W**2,[0],M,M), sp.spdiags(eps_vec,[0],N,N))
+        entries_diag, indices_diag = get_entries_indices(D)
+
+        entries_a = npa.hstack((entries_diag, entries_c))
+        indices_a = npa.hstack((indices_diag, indices_c))
+
+        Nfreq = npa.shape(delta_matrix)[0]
+        for k in npa.arange(Nfreq):
```

Comment:
Could you put a comment in here linking to a paper (jerry's?) that defines these matrices?  Just for reference.

Reply:
Done!